### DISCUSSION TITLE
OIDC Authentication

### DISCUSSION THREAD

**Discussion by TobiasGrether** [outline/outline#3478] [Self Hosting]
I've barely found any information on OIDC authentication, but I saw that a few environmental variables exist for it. Is there any document on it and is OIDC auth available in any version of outline? (Of course talking selfhosted here)

**Comment by IkerSaint** [comment]
I was able to ingrate Outline with Authentik using this env variable:

```
OIDC_CLIENT_ID=<yousupersecretclientid>
OIDC_CLIENT_SECRET=<evenmoresecretsecret>
OIDC_AUTH_URI=https://<yourauthdomain>/application/o/authorize/
OIDC_TOKEN_URI=https://<yourauthdomain>/application/o/token/
OIDC_USERINFO_URI=https://<yourauthdomain>/application/o/userinfo/
OIDC_USERNAME_CLAIM=email
OIDC_DISPLAY_NAME=openid
OIDC_SCOPES="openid profile email"
FORCE_HTTPS=true # this one is very important
URL=https://yourexternalurl # Also important
```

Additionally, on the Authentik side, the creation of the provider and application is quite easy, the redirect URL is https://yourexternalur/auth/oidc.callback

**Comment by gnowland** [comment]
What's the best way to debug OIDC auth? I've added the above environment variables but upon redirect from Keycloak I'm receiving the following error:
```txt
Authentication failed – we were unable to sign you in at this time. Please try again.
```
And my server logs indicate:
```txt
{
  "error": "Failed to obtain access token",
  "stack": "InternalOAuthError: Failed to obtain access token
            at OAuth2Strategy._createOAuthError (/opt/outline/node_modules/passport-oauth2/lib/strategy.js:423:17)
            at /opt/outline/node_modules/passport-oauth2/lib/strategy.js:177:45
            at /opt/outline/node_modules/oauth/lib/oauth2.js:191:18
            at ClientRequest.<anonymous> (/opt/outline/node_modules/oauth/lib/oauth2.js:162:5)
            at ClientRequest.emit (node:events:526:28)
            at ClientRequest.emit (node:domain:475:12)
            at TLSSocket.socketErrorListener (node:_http_client:442:9)
            at TLSSocket.emit (node:events:526:28)
            at TLSSocket.emit (node:domain:475:12)
            at emitErrorNT (node:internal/streams/destroy:157:8)
            at emitErrorCloseNT (node:internal/streams/destroy:122:3)
            at processTicksAndRejections (node:internal/process/task_queues:83:21)",
  "level": "error",
  "message": "Error during authentication"
}
```

**Reply by 1itt1eB0y** [reply]
check your keycloak API address, maybe there is wrong setting.

**Comment by 1itt1eB0y** [comment]
You can try authlia, authentic, keycloak, dex and other more.
Set up a OIDC server first, create a client for outline, and you will get the API address. Fill the configuration with these addresses, then you can use OIDC to login.

**Reply by gnowland** [reply]
Thanks I think there must be a misconfiguration of the keycloak server I’ll contact my keycloak admin

**Reply by chaishuanzhu** [reply]
authlia
```
{"label":"http","level":"info","message":"  <-- GET /auth/oidc.callback?code=eMJUDzgRHuebKye9ZP3SfX6rZlY6FNQ_3NricyBvzV0.YvkFnCr8pGcMrVzmqMJAORyS7BIs1PuS9LQHNBte-Oo&scope=openid&state=059af5ac501c5b39"}

{"label":"http","level":"info","message":"  <-- GET /"}

{"label":"http","level":"info","message":"  --> GET / 200 10ms 2.08kb"}

{"error":"Failed to obtain access token","stack":"Error: Failed to obtain access token\n    at /opt/outline/node_modules/passport-oauth2/lib/strategy.js:178:49\n    at /opt/outline/node_modules/oauth/lib/oauth2.js:209:7\n    at passBackControl (/opt/outline/node_modules/oauth/lib/oauth2.js:134:9)\n    at IncomingMessage.<anonymous> (/opt/outline/node_modules/oauth/lib/oauth2.js:157:7)\n    at IncomingMessage.emit (node:events:538:35)\n    at IncomingMessage.emit (node:domain:475:12)\n    at endReadableNT (node:internal/streams/readable:1345:12)\n    at processTicksAndRejections (node:internal/process/task_queues:83:21)","level":"error","message":"Error during authentication"}

{"label":"http","level":"info","message":"  --> POST /auth/oidc.callback?code=eMJUDzgRHuebKye9ZP3SfX6rZlY6FNQ_3NricyBvzV0.YvkFnCr8pGcMrVzmqMJAORyS7BIs1PuS9LQHNBte-Oo&scope=openid&state=059af5ac501c5b39 302 830ms 69b"}
```

**Reply by Wlmrsi** [reply]
I have the same problem with Authentik, and can not find a way to resolve it...

**Reply by reejosam** [reply]
Getting same error with a custom django oidc server that works with other apps. I suspect it might have something do with not using https being an internal server. Information on the error would be helpful to debug.

**Reply by tommoor** [reply]
It's a limitation of the underlying library unfortunately, we'd have to fork `passport` to get better errors here… maybe it'd be worth it.

> I suspect it might have something do with not using https being an internal server

For authentication you should _always_ use https, it's expected that all parts of Outline run with SSL enabled.

**Reply by felixsanz** [reply]
> It's a limitation of the underlying library unfortunately, we'd have to fork `passport` to get better errors here… maybe it'd be worth it.
> 
> > I suspect it might have something do with not using https being an internal server
> 
> For authentication you should _always_ use https, it's expected that all parts of Outline run with SSL enabled.

But what if I want to do SSL termination at reverse proxy level? The Outline<>Proxy connection should be under SSL too?

**Reply by almereyda** [reply]
That appears for you to decide, as it is an intermediary transmission that does not provide its own endpoint which would be used by third parties.
